library/std/src/collections/mod.rs

```diff
@@ -194,8 +194,9 @@
 //! at 0. For ordered collections like [`BTreeMap`], this means that the items
 //! will be yielded in sorted order. For unordered collections like [`HashMap`],
 //! the items will be yielded in whatever order the internal representation made
-//! most convenient. This is great for reading through all the contents of the
-//! collection.
+//! most convenient. This order is guaranteed to be the same across multiple calls
+//! to the same collection, as long as the collection isn't modified in between.
```

Comment:
"modified" is vague. E.g. a call to `reserve()` does not modify its contents but still changes iteration order.
On the other hand `.values_mut()` currently doesn't restructure a hashmap.

Reply:
I agree, maybe "as long as the internal state of the collection isn't modified" would be better?

Reply:
Internal state is implementation detail and hidden for user, so there no way to knew when modified or not.

Reply:
I think we can use `&self` as conservative approximation for non-modifying access.

Reply:
I agree that this wording is unclear: It's easy to argue that `.reserve()` does not "modify the collection", but it can still change the iteration order.

> I think we can use `&self` as conservative approximation for non-modifying access.

`.values_mut()` takes an `&mut self`, but doesn't change the iteration order.

Maybe something like: "[..] as long as no element or capacity is added or removed."?